R/parse_variable_indices.R

```diff
@@ -0,0 +1,148 @@
+  })
+  variable_indices_info <- lapply(var_names, function (x) {
+    # dimensions counts the number of unique values of each index.
+    # implied_dimensions gives the maximum for integer indices with no values <= 0
+    # indices is a data frame with one line per combination of indices, including implied indices
+    # position gives the position in the original vector of variables (passed to `parse_variable_indices`)
+    # where a given combination of indices is found (NA if the comination does not exist)
+    indices_info <- named_list(c("dimensionality", "dimensions", "implied_dimensions", "index_names", "indices", "position"))
+    indices_info$dimensionality <- dimensionality[[x]]
+    var_i <- vars == x
+    var_length <- sum(var_i)
+    if (dimensionality[x] == 0) {
+      # single variable, no indices
+      indices_info$dimensions <- indices_info$implied_dimensions <- NA
+      indices_info$indices <- as.integer(1)
+      dimnames(indices_info$indices) <- NULL
+      indices_info$implied_dimensions <- NA
+      indices_info$internal_position <- 1
+    } else {
+      indices <- sapply(vars_indices[var_i], `[[`, 2)
+      indices <- as.data.frame(do.call(rbind, strsplit(indices, ",")),
+                               stringsAsFactors = FALSE)
+      indices_info$dimensions <- unname(apply(indices, 2, function(x){length(unique(x))}))
+      unique_indices <- vector("list", length(indices))
+      .dimnames <- vector("list", length(indices))
+      names(unique_indices) <- names(indices)
+      for (i in seq_along(indices)) {
+        numeric_index <- suppressWarnings(as.numeric(indices[[i]]))
+        if (!anyNA(numeric_index) && rlang::is_integerish(numeric_index)) {
+          # for integer indices, we need to convert them to integers
+          # so that we can sort them in numerical order (not string order)
+          if (min(numeric_index) >= 1) {
+            # integer indices >= 1 are forced to lower bound of 1 + no dimnames
+            indices[[i]] <- as.integer(numeric_index)
+            unique_indices[[i]] <- seq.int(1, max(numeric_index))
+          } else {
+            # indices with values < 1 are filled in between the min and max
+            indices[[i]] <- numeric_index
+            unique_indices[[i]] <- seq.int(min(numeric_index), max(numeric_index))
+          }
+        } else {
+          # we convert non-numeric indices to factors so that we can force them
+          # to be ordered as they appear in the data (rather than in alphabetical order)
+          factor_levels <- unique(indices[[i]])
+          indices[[i]] <- factor(indices[[i]], levels = factor_levels)
+          # these aren't sorted so they appear in original order
+          unique_indices[[i]] <- factor(factor_levels, levels = factor_levels)
+          .dimnames[[i]] <- unique_indices[[i]]
+        }
+      }
+      
+      indices_info$index_names <- .dimnames
+      indices_info$implied_dimensions <- unname(sapply(unique_indices, length))
```

Comment:
As above, I would use `lengths()` here
```suggestion
      indices_info$implied_dimensions <- unname(lengths(unique_indices))
```